core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java

```diff
@@ -157,170 +155,170 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
 
   @Override
   public CollectionBuilder itemMediaType(MediaType mediaType) {
-    if (dataType instanceof DefaultCollectionDataType) {
-      dataType =
-          ((DataTypeBuilder.DataTypeCollectionTypeBuilder) DataType.builder(this.dataType)).itemMediaType(mediaType).build();
+    if (payload.getDataType() instanceof DefaultCollectionDataType) {
+      payload = new TypedValue(payload.getValue(),
+                               ((DataTypeBuilder.DataTypeCollectionTypeBuilder) builder(payload.getDataType()))
+                                   .itemMediaType(mediaType).build());
     } else {
       throw new IllegalStateException("Item MediaType cannot be set, because payload is not a collection");
     }
     return this;
   }
 
   @Override
-  public InternalMessage.Builder attributes(TypedValue<?> typedValue) {
-    this.attributes = typedValue.getValue();
-    this.attributesDataType = typedValue.getDataType();
+  public InternalMessage.CollectionBuilder attributes(TypedValue<?> attributes) {
```

Comment:
DW does reflection, so it's not limited by this since it will be dealing with the implementation that's all builders at once.